Use NSStepperCell to draw spin buttons rather than CoreUI

lilyspiniolas opened the pull request:
#### f5e8aa3294f74adc9adc3d99cc5c2724e6bd86ba
<pre>
Use NSStepperCell to draw spin buttons rather than CoreUI
<a href="https://bugs.webkit.org/show_bug.cgi?id=284655">https://bugs.webkit.org/show_bug.cgi?id=284655</a>
<a href="https://rdar.apple.com/140298961">rdar://140298961</a>

Reviewed by Aditya Keerthi.

NSStepperCell is now used to draw spin buttons when its property
&apos;incrementing&apos; is available. When the property is unavailable,
the original CoreUI implementation is used instead.

* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/WebCore/PAL/pal/spi/cocoa/NSStepperCellSPI.h: Added.
* Source/WebCore/platform/graphics/mac/controls/ControlFactoryMac.h:
* Source/WebCore/platform/graphics/mac/controls/ControlFactoryMac.mm:
(WebCore::ControlFactoryMac::stepperCell const):
(WebCore::ControlFactoryMac::createPlatformInnerSpinButton):
* Source/WebCore/platform/graphics/mac/controls/InnerSpinButtonMac.h:
* Source/WebCore/platform/graphics/mac/controls/InnerSpinButtonMac.mm:
(WebCore::InnerSpinButtonMac::InnerSpinButtonMac):
(WebCore::InnerSpinButtonMac::cellSize const):
(WebCore::InnerSpinButtonMac::draw):
(WebCore::InnerSpinButtonMac::drawWithCell):
(WebCore::InnerSpinButtonMac::drawWithCoreUI):
(WebCore::InnerSpinButtonMac::rectForBounds const):
(WebCore::InnerSpinButtonMac::updateCellStates):
* Source/WebCore/platform/mac/ThemeMac.mm:
(WebCore::stepperSizes):

Canonical link: <a href="https://commits.webkit.org/288070@main">https://commits.webkit.org/288070@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a2546478027b197e395dea0a9f71136575ff245c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/81816 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/1342 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/35770 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/86363 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/32812 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/83922 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/1375 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/9163 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/86363 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/21534 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/84886 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/1003 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/74454 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/86363 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/902 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/28629 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/31265 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/72274 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/29236 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/87799 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/9056 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/6430 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/72154 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/9241 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/70275 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/71386 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17781 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/15483 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/14406 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/9007 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/14539 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/8848 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/12371 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/10656 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/platform/graphics/mac/controls/InnerSpinButtonMac.mm:
```diff
@@ -43,13 +44,25 @@
 IntSize InnerSpinButtonMac::cellSize(NSControlSize controlSize, const ControlStyle&) const
 {
+#if HAVE(NSSTEPPERCELL_INCREMENTING)
+    if ([m_stepperCell respondsToSelector:@selector(setIncrementing:)]) {
```
I find it nicer to use `instancesRespondToSelector:` in cases, where you're not actually using the instance.

pxlcoder commented on Source/WebCore/platform/graphics/mac/controls/InnerSpinButtonMac.mm:
```diff
@@ -59,7 +72,35 @@
+void InnerSpinButtonMac::draw(GraphicsContext& context, const FloatRoundedRect& borderRect, float deviceScaleFactor, const ControlStyle& style)
+{
+#if HAVE(NSSTEPPERCELL_INCREMENTING)
+    if ([m_stepperCell respondsToSelector:@selector(setIncrementing:)]) {
```
Same comment about `instancesRespondToSelector:`.

pxlcoder commented on Source/WebCore/platform/graphics/mac/controls/InnerSpinButtonMac.mm:
```diff
@@ -107,6 +148,43 @@
+
+    const auto& states = style.states;
+
+    m_stepperCell.get().incrementing = states.contains(ControlStyle::State::SpinUp);
```
```suggestion
    [m_stepperCell setIncrementing:states.contains(ControlStyle::State::SpinUp)];
```

pxlcoder commented on Source/WebCore/platform/graphics/mac/controls/InnerSpinButtonMac.mm:
```diff
@@ -43,13 +44,25 @@
 {
+#if HAVE(NSSTEPPERCELL_INCREMENTING)
+    if ([m_stepperCell respondsToSelector:@selector(setIncrementing:)]) {
+        static const IntSize sizes[] = {
```
We probably shouldn't specify a large size, given one doesn't exist. I think it will cause problems with `controlSizeForSize` as-is. We may also need to teach that method to ignore controls that don't support a large size.

pxlcoder replied:
Oh, I see you force `NSControlSizeRegular` below, so maybe okay?

pxlcoder commented on Source/WebCore/platform/graphics/mac/controls/ControlFactoryMac.mm:
```diff
@@ -214,6 +214,14 @@
+{
+    if (!m_stepperCell)
+        m_stepperCell = adoptNS([[NSStepperCell alloc] init]);
+
```
Nit: Remove newline to match the other methods in this file.

pxlcoder commented on Source/WebCore/platform/graphics/mac/controls/InnerSpinButtonMac.h:
```diff
@@ -37,12 +37,21 @@ class InnerSpinButtonPart;
     WTF_MAKE_TZONE_ALLOCATED(InnerSpinButtonMac);
 public:
-    InnerSpinButtonMac(InnerSpinButtonPart&, ControlFactoryMac&);
+    InnerSpinButtonMac(InnerSpinButtonPart&, ControlFactoryMac&, NSStepperCell*);
```
```suggestion
    InnerSpinButtonMac(InnerSpinButtonPart&, ControlFactoryMac&, NSStepperCell *);
```

pxlcoder commented on Source/WebCore/platform/graphics/mac/controls/InnerSpinButtonMac.mm:
```diff
@@ -43,13 +44,25 @@
+#if HAVE(NSSTEPPERCELL_INCREMENTING)
+    if ([m_stepperCell respondsToSelector:@selector(setIncrementing:)]) {
+        static const IntSize sizes[] = {
+            { 19, 28 },
+            { 15, 22 },
+            { 13, 18 },
+            { 19, 28 }
```
I see these metrics are slightly different from the existing. We also need to update `stepperSizes()` in `ThemeMac`.

We really need to have a single source of truth, and then also make these programmatically determined, but not asking for that in this patch 🙂

pxlcoder commented on Source/WebCore/platform/mac/ThemeMac.mm:
```diff
@@ -421,6 +433,10 @@ static LengthSize switchSize(const LengthSize& zoomedSize, float zoomFactor)
     case StyleAppearance::InnerSpinButton: {
         static const int stepperMargin[4] = { 0, 0, 0, 0 };
```
I think you also need to add `stepperMargins` to match the cell outsets. Otherwise, I would expect issues (painting artifacts) on repaint.

pxlcoder commented on Source/WebCore/platform/mac/ThemeMac.mm:
```diff
@@ -214,8 +215,33 @@ static LengthSize radioSize(const LengthSize& zoomedSize, float zoomFactor)
+        return margins[controlSize];
+    }
+#endif
+    UNUSED_PARAM(controlSize);
```
Nit:
```suggestion
#else
    UNUSED_PARAM(controlSize);
#endif
```

lilyspiniolas committed:
Use NSStepperCell to draw spin buttons rather than CoreUI
https://bugs.webkit.org/show_bug.cgi?id=284655
rdar://140298961

Reviewed by Aditya Keerthi.

NSStepperCell is now used to draw spin buttons when its property
'incrementing' is available. When the property is unavailable,
the original CoreUI implementation is used instead.

* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/WebCore/PAL/pal/spi/cocoa/NSStepperCellSPI.h: Added.
* Source/WebCore/platform/graphics/mac/controls/ControlFactoryMac.h:
* Source/WebCore/platform/graphics/mac/controls/ControlFactoryMac.mm:
(WebCore::ControlFactoryMac::stepperCell const):
(WebCore::ControlFactoryMac::createPlatformInnerSpinButton):
* Source/WebCore/platform/graphics/mac/controls/InnerSpinButtonMac.h:
* Source/WebCore/platform/graphics/mac/controls/InnerSpinButtonMac.mm:
(WebCore::InnerSpinButtonMac::InnerSpinButtonMac):
(WebCore::InnerSpinButtonMac::cellSize const):
(WebCore::InnerSpinButtonMac::draw):
(WebCore::InnerSpinButtonMac::drawWithCell):
(WebCore::InnerSpinButtonMac::drawWithCoreUI):
(WebCore::InnerSpinButtonMac::rectForBounds const):
(WebCore::InnerSpinButtonMac::updateCellStates):
* Source/WebCore/platform/mac/ThemeMac.mm:
(WebCore::stepperSizes):

Canonical link: https://commits.webkit.org/288070@main
